File: ios/MenCare/Features/Admin/Manage/ServiceEditScreen.swift
import SwiftUI
import Observation

@MainActor
@Observable
final class ServiceEditViewModel {

    private let catalog: CatalogRepository
    private let serviceId: String?

    var isNew: Bool { serviceId == nil }
    var name = ""
    var durationMinutes = 30
    var priceEuros = ""
    var enabledOperatorIds: Set<String> = []
    var saved = false
    var saving = false
    /// Chiave del campo → messaggio: prima il tasto "Salva" con il nome vuoto
    /// semplicemente non faceva nulla, senza dire perché.
    var fieldErrors: [String: String] = [:]
    var saveError: String?

    @ObservationIgnored private var submitted = false

    let state = LoadState()
    private(set) var operators: [Operator] = []
    private var editing: Service?

    init(catalog: CatalogRepository, serviceId: String?) {
        self.catalog = catalog
        self.serviceId = serviceId
    }

    func load() async {
        await state.run {
            let snapshot = try await catalog.catalog()
            operators = snapshot.operators
            let service = serviceId.flatMap { id in snapshot.services.first { $0.id == id } }
            editing = service
            if let service {
                name = service.name
                durationMinutes = service.durationMinutes
                priceEuros = formatCentsAsInput(service.priceCents)
                enabledOperatorIds = Set(snapshot.operators.filter { $0.serviceIds.contains(service.id) }.map(\.id))
            }
        }
    }

    func fieldChanged() {
        saveError = nil
        saved = false
        if submitted { fieldErrors = validate() }
    }

    private func validate() -> [String: String] {
        var errors: [String: String] = [:]
        // Il nome di un servizio non è un nome proprio: "Shampoo + taglio" deve passare.
        errors["name"] = validateRequiredText(name).message
        errors["duration"] = validateDuration(durationMinutes).message
        errors["price"] = validatePriceInput(priceEuros).message
        return errors.compactMapValues { $0 }
    }

    func toggleOperator(_ operatorId: String) {
        if enabledOperatorIds.contains(operatorId) {
            enabledOperatorIds.remove(operatorId)
        } else {
            enabledOperatorIds.insert(operatorId)
        }
    }

    func save() {
        guard !saving else { return }
        submitted = true
        let errors = validate()
        if !errors.isEmpty {
            fieldErrors = errors
            saveError = L("validation_form_invalid")
            return
        }
        // `validatePrice` ha già accettato il testo, quindi la conversione non
        // può fallire; si arrotonda ai centesimi, non si tronca.
        guard let cents = parsePriceToCents(priceEuros) else {
            fieldErrors = ["price": L("validation_price_invalid")]
            return
        }
        fieldErrors = [:]
        saveError = nil
        saving = true
        Task {
            // Servizio e abilitazioni in un colpo solo: il server ha una rotta
            // apposta, cosi' non resta un listino a meta' se una scrittura salta.
            let result = await catalog.saveService(
                Service(
                    id: serviceId ?? "",
                    name: name.trimmingCharacters(in: .whitespaces),
                    durationMinutes: durationMinutes,
                    priceCents: cents,
                    description: editing?.description,
                    featured: editing?.featured ?? false
                ),
                operatorIds: enabledOperatorIds
            )
            saving = false
            switch result {
            case .success:
                saved = true
            case .failure(let error):
                saveError = error.displayMessage
            }
        }
    }
}

struct ServiceEditScreen: View {
    @Bindable var viewModel: ServiceEditViewModel
    let onBack: () -> Void

    var body: some View {
        VStack(spacing: 0) {
            DarkHeader(contentPadding: EdgeInsets(top: 0, leading: 0, bottom: 8, trailing: 0)) {
                BrandTopBar(
                    title: L(viewModel.isNew ? "svc_new_title" : "svc_edit_title"),
                    onBack: onBack,
                    backLabel: L("manage_tab_services")
                ) {
                    BarAction(text: L("svc_save_short"), action: viewModel.save, color: .oliveLight)
                }
            }

            ScrollView {
                VStack(alignment: .leading, spacing: 14) {
                    FilledTextField(
                        label: L("svc_name"), text: $viewModel.name,
                        autocapitalization: .sentences,
                        error: viewModel.fieldErrors["name"], outlined: true
                    )
                    .onChange(of: viewModel.name) { viewModel.fieldChanged() }

                    HStack(alignment: .top, spacing: 10) {
                        DurationField(
                            label: L("svc_duration"),
                            minutes: $viewModel.durationMinutes,
                            error: viewModel.fieldErrors["duration"]
                        )
                        .onChange(of: viewModel.durationMinutes) { viewModel.fieldChanged() }
                        VStack(alignment: .leading, spacing: 8) {
                            BrandSectionLabel(text: L("svc_price"))
                            PriceField(
                                label: "", text: $viewModel.priceEuros,
                                error: viewModel.fieldErrors["price"]
                            )
                            .onChange(of: viewModel.priceEuros) { viewModel.fieldChanged() }
                        }
                    }

                    FormErrorBanner(message: viewModel.saveError)

                    VStack(alignment: .leading, spacing: 8) {
                        BrandSectionLabel(text: L("svc_operators"))
                        VStack(spacing: 0) {
                            ForEach(Array(viewModel.operators.enumerated()), id: \.element.id) { index, op in
                                if index > 0 {
                                    Rectangle().fill(Color.stoneBorder).frame(height: 1)
                                }
                                Button {
                                    viewModel.toggleOperator(op.id)
                                } label: {
                                    HStack(spacing: 11) {
                                        Circle()
                                            .fill(Color.bone)
                                            .frame(width: 32, height: 32)
                                            .overlay(
                                                Text(op.initials)
                                                    .font(Typo.cormorant(12, weight: .regular))
                                                    .foregroundStyle(Color.ink)
                                            )
                                        Text(op.name)
                                            .font(Typo.bodyLarge)
                                            .foregroundStyle(Color.ink)
                                        Spacer()
                                        let enabled = viewModel.enabledOperatorIds.contains(op.id)
                                        RoundedRectangle(cornerRadius: 6)
                                            .fill(enabled ? Color.oliveWood : Color.bone)
                                            .frame(width: 24, height: 24)
                                            .overlay(
                                                enabled
                                                    ? Image(systemName: "checkmark")
                                                        .font(.system(size: 12, weight: .semibold))
                                                        .foregroundStyle(Color.bone)
                                                    : nil
                                            )
                                    }
                                    .padding(.horizontal, 14)
                                    .padding(.vertical, 12)
                                }
                                .buttonStyle(.plain)
                            }
                        }
                        .background(RoundedRectangle(cornerRadius: 16).fill(Color.stone))
                        .clipShape(RoundedRectangle(cornerRadius: 16))
                    }
                }
                .padding(.horizontal, 20)
                .padding(.top, 18)
                .padding(.bottom, 16)
                .readableWidth()
            }

            BottomActionBar {
                AccentButton(
                    text: L(viewModel.isNew ? "svc_save" : "svc_save_changes"),
                    action: viewModel.save,
                    loading: viewModel.saving,
                    height: 56,
                    corner: 16
                )
                .padding(.bottom, 10)
                .readableWidth()
            }
        }
        .background(Color.bone)
        .task { await viewModel.load() }
        .onChange(of: viewModel.saved) {
            if viewModel.saved { onBack() }
        }
    }

}

File: android/feature/admin/src/main/kotlin/com/devora/mencare/feature/admin/manage/ServiceEditScreen.kt
package com.devora.mencare.feature.admin.manage

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.imePadding
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.Check
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.devora.mencare.core.designsystem.component.AccentButton
import com.devora.mencare.core.designsystem.component.BarAction
import com.devora.mencare.core.designsystem.component.BottomActionBar
import com.devora.mencare.core.designsystem.component.BrandSectionLabel
import com.devora.mencare.core.designsystem.component.BrandTopBar
import com.devora.mencare.core.designsystem.component.DarkHeader
import com.devora.mencare.core.designsystem.component.DurationField
import com.devora.mencare.core.designsystem.component.FilledTextField
import com.devora.mencare.core.designsystem.component.PriceField
import com.devora.mencare.core.designsystem.component.readableWidth
import com.devora.mencare.core.designsystem.component.validationMessageOrNull
import com.devora.mencare.core.designsystem.theme.Bone
import com.devora.mencare.core.designsystem.theme.Cormorant
import com.devora.mencare.core.designsystem.theme.ErrorRed
import com.devora.mencare.core.designsystem.theme.Ink
import com.devora.mencare.core.designsystem.theme.OliveLight
import com.devora.mencare.core.designsystem.theme.OliveWood
import com.devora.mencare.core.designsystem.theme.Radii
import com.devora.mencare.core.designsystem.theme.StoneBorder
import com.devora.mencare.feature.admin.R

@Composable
fun ServiceEditScreen(
    onBack: () -> Unit,
    viewModel: ServiceEditViewModel = hiltViewModel(),
) {
    val state by viewModel.state.collectAsStateWithLifecycle()

    LaunchedEffect(state.saved) {
        if (state.saved) onBack()
    }

    Column(modifier = Modifier.fillMaxSize().background(Bone)) {
        DarkHeader(
            contentPadding = PaddingValues(start = 0.dp, end = 0.dp, top = 0.dp, bottom = 8.dp),
        ) {
            BrandTopBar(
                title = stringResource(
                    if (state.isNew) R.string.svc_new_title else R.string.svc_edit_title,
                ),
                onBack = onBack,
                backLabel = stringResource(R.string.manage_tab_services),
                trailing = { BarAction(stringResource(R.string.svc_save_short), viewModel::save, color = OliveLight) },
            )
        }

        Column(
            modifier = Modifier
                .weight(1f)
                .verticalScroll(rememberScrollState())
                .readableWidth()
                .imePadding()
                .padding(horizontal = 20.dp)
                .padding(top = 18.dp, bottom = 16.dp),
            verticalArrangement = Arrangement.spacedBy(14.dp),
        ) {
            FilledTextField(
                value = state.name,
                onValueChange = viewModel::setName,
                label = stringResource(R.string.svc_name),
                error = validationMessageOrNull(state.nameError),
            )

            Row(horizontalArrangement = Arrangement.spacedBy(10.dp)) {
                Column(Modifier.weight(1f)) {
                    BrandSectionLabel(stringResource(R.string.svc_duration))
                    Spacer(Modifier.height(8.dp))
                    DurationField(
                        minutes = state.durationMinutes,
                        onMinutesChange = viewModel::setDuration,
                        error = state.durationError,
                    )
                }
                Column(Modifier.weight(1f)) {
                    BrandSectionLabel(stringResource(R.string.svc_price))
                    Spacer(Modifier.height(8.dp))
                    PriceField(
                        value = state.priceEuros,
                        onValueChange = viewModel::setPrice,
                        label = "",
                        error = state.priceError,
                        height = 56.dp,
                    )
                }
            }

            if (state.saveFailed) {
                Text(
                    stringResource(R.string.svc_error_save),
                    style = MaterialTheme.typography.bodySmall,
                    color = ErrorRed,
                )
            }

            Column {
                BrandSectionLabel(stringResource(R.string.svc_operators))
                Spacer(Modifier.height(8.dp))
                Column(
                    modifier = Modifier.fillMaxWidth().clip(Radii.Md).background(Bone).border(1.5.dp, StoneBorder, Radii.Md),
                ) {
                    state.operators.forEachIndexed { index, operator ->
                        if (index > 0) HorizontalDivider(color = StoneBorder)
                        Row(
                            modifier = Modifier
                                .fillMaxWidth()
                                .clickable { viewModel.toggleOperator(operator.id) }
                                .padding(horizontal = 14.dp, vertical = 12.dp),
                            verticalAlignment = Alignment.CenterVertically,
                        ) {
                            Box(
                                modifier = Modifier.size(32.dp).clip(CircleShape).background(Bone),
                                contentAlignment = Alignment.Center,
                            ) {
                                Text(operator.initials, fontFamily = Cormorant, fontSize = 12.sp, color = Ink)
                            }
                            Text(
                                operator.name,
                                style = MaterialTheme.typography.bodyLarge,
                                color = Ink,
                                modifier = Modifier.weight(1f).padding(start = 11.dp),
                            )
                            val enabled = operator.id in state.enabledOperatorIds
                            Box(
                                modifier = Modifier
                                    .size(24.dp)
                                    .clip(RoundedCornerShape(6.dp))
                                    .background(if (enabled) OliveWood else Bone),
                                contentAlignment = Alignment.Center,
                            ) {
                                if (enabled) {
                                    Icon(
                                        Icons.Outlined.Check,
                                        contentDescription = null,
                                        tint = Bone,
                                        modifier = Modifier.size(15.dp),
                                    )
                                }
                            }
                        }
                    }
                }
            }
        }

        BottomActionBar(modifier = Modifier.navigationBarsPadding()) {
            AccentButton(
                modifier = Modifier.readableWidth(),
                text = stringResource(
                    if (state.isNew) R.string.svc_save else R.string.svc_save_changes,
                ),
                onClick = viewModel::save,
                loading = state.saving,
                height = 56.dp,
                shape = RoundedCornerShape(16.dp),
            )
            Spacer(Modifier.height(10.dp))
        }
    }
}
